substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -94,17 +94,246 @@ private PosixJavaNetSubstitutions() {
+        //        }
+        if (lport == 0) {
+            if (VmPrimsJVM.JVM_GetSockName(fd, him, len_Pointer) == -1) {
+                throw new SocketException(PosixUtils.lastErrorString("Error getting socket name"));
+            }
+            // 602         localport = NET_GetPortFromSockaddr((struct sockaddr *)&him);
+            lport = JavaNetNetUtilMD.NET_GetPortFromSockaddr(him);
+            // 603         (*env)->SetIntField(env, this, psi_localportID, localport);
+            Util_java_net_PlainDatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).localPort = lport;
+        } else {
+            // 605         (*env)->SetIntField(env, this, psi_localportID, localport);
+            Util_java_net_PlainDatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).localPort = lport;
+        }
     }
 
     @Substitute
     @SuppressWarnings({"static-method", "unused"})
-    protected void send(DatagramPacket p) throws IOException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.PlainDatagramSocketImpl.send(DatagramPacket)");
+    protected void send(DatagramPacket packet) throws IOException {
+        //    char BUF[MAX_BUFFER_LEN];
+        //    char *fullPacket = NULL;
+        //        int ret, mallocedPacket = JNI_FALSE;
+        //    /* The object's field */
+        //        jobject fdObj = (*env)->GetObjectField(env, this, pdsi_fdID);
+        //        jint trafficClass = (*env)->GetIntField(env, this, pdsi_trafficClassID);
+        //
+        //        jbyteArray packetBuffer;
+        //        jobject packetAddress;
+        //        jint packetBufferOffset, packetBufferLen, packetPort;
+        //        jboolean connected;
+        //
+        //    /* The fdObj'fd */
+        //        jint fd;
+        FileDescriptor fdObj = Util_java_net_DatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).fd;
+        //
+        //        SOCKADDR rmtaddr, *rmtaddrP=&rmtaddr;
+        //        int len;
+        //
+        //        if (IS_NULL(fdObj)) {
+        if (fdObj == null) {
+            //            JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException",
+            //                            "Socket closed");
+            throw new SocketException("Socket closed");
+        }
+        //        fd = (*env)->GetIntField(env, fdObj, IO_fd_fdID);
+        int fd = PosixUtils.getFD(fdObj);
+        //
+        //        if (IS_NULL(packet)) {
+        if (packet == null) {
+            //            JNU_ThrowNullPointerException(env, "packet");
+            throw new NullPointerException("packet");
+        }
+        //        packetBuffer = (*env)->GetObjectField(env, packet, dp_bufID);
+        //        packetAddress = (*env)->GetObjectField(env, packet, dp_addressID);
+        //        if (IS_NULL(packetBuffer) || IS_NULL(packetAddress)) {
+        if (packet.getData() == null || packet.getAddress() == null) {
```

Comment:
Whose mistake (!) is it that this field is accessed with a method called `getData()`, when `DatagramPacket` does not seem to have a field named `data`?  It has a field called `but`, but that should be accessed by turning `packet` into a `Target_java_net_DatagramPacket` and looking that the `buf` field.  Oh, I see, you are just using the Java access methods.  I guess that's allowed.  Might be worth a comment in this rare case where there are public accessors for the fields.

Hmm, the Java access methods are all `synchronized`.  Note that the C code doesn't obey that protocol.  In particular, if the C code is reading the length and the data while someone else (even in synchronized Java code) is calling `setData` which updates a bunch of the fields, supposedly atomically.  Of course, Java code (or this code) that separately calls `getData` and `getLength` could get mis-matched values, too.  Turn over a rock, find a bug.  Maybe the worst that can happen is that the Java code gets an `ArrayIndexOutOfBoundsException`.  Maybe the C code could get itself into more trouble.

What happens if you call a synchronized access method and block?  Might you have just introduced a deadlock, since no one is expecting this method to block on the field accesses? I *think* not, because the accesses should all complete without any long-lasting contention.  But I am sure I do not want to debug that.